// bpf/dist/exception.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkStatus = exports.BPFError = exports.libbpfErrnoMessages = exports.LibbpfErrno = void 0;
const util = __importStar(require("util"));
const { getSystemErrorName } = util;
/**
 * libbpf-specific errno codes
 *
 * Keep synchronized with `deps/libbpf/src/libbpf.h`.
 */
var LibbpfErrno;
(function (LibbpfErrno) {
    /** Something wrong in libelf */
    LibbpfErrno[LibbpfErrno["LIBELF"] = 4000] = "LIBELF";
    /** BPF object format invalid */
    LibbpfErrno[LibbpfErrno["FORMAT"] = 4001] = "FORMAT";
    /** Incorrect or no 'version' section */
    LibbpfErrno[LibbpfErrno["KVERSION"] = 4002] = "KVERSION";
    /** Endian mismatch */
    LibbpfErrno[LibbpfErrno["ENDIAN"] = 4003] = "ENDIAN";
    /** Internal error in libbpf */
    LibbpfErrno[LibbpfErrno["INTERNAL"] = 4004] = "INTERNAL";
    /** Relocation failed */
    LibbpfErrno[LibbpfErrno["RELOC"] = 4005] = "RELOC";
    /** Load program failure for unknown reason */
    LibbpfErrno[LibbpfErrno["LOAD"] = 4006] = "LOAD";
    /** Kernel verifier blocks program loading */
    LibbpfErrno[LibbpfErrno["VERIFY"] = 4007] = "VERIFY";
    /** Program too big */
    LibbpfErrno[LibbpfErrno["PROG2BIG"] = 4008] = "PROG2BIG";
    /** Incorrect kernel version */
    LibbpfErrno[LibbpfErrno["KVER"] = 4009] = "KVER";
    /** Kernel doesn't support this program type */
    LibbpfErrno[LibbpfErrno["PROGTYPE"] = 4010] = "PROGTYPE";
    /** Wrong pid in netlink message */
    LibbpfErrno[LibbpfErrno["WRNGPID"] = 4011] = "WRNGPID";
    /** Invalid netlink sequence */
    LibbpfErrno[LibbpfErrno["INVSEQ"] = 4012] = "INVSEQ";
    /** netlink parsing error */
    LibbpfErrno[LibbpfErrno["NLPARSE"] = 4013] = "NLPARSE";
})(LibbpfErrno = exports.LibbpfErrno || (exports.LibbpfErrno = {}));
/**
 * mapping of libbpf-specific errno codes to messages
 *
 * Keep synchronized with `deps/libbpf/src/libbpf.h`.
 */
exports.libbpfErrnoMessages = {
    [LibbpfErrno.LIBELF]: 'Something wrong in libelf',
    [LibbpfErrno.FORMAT]: 'BPF object format invalid',
    [LibbpfErrno.KVERSION]: "Incorrect or no 'version' section",
    [LibbpfErrno.ENDIAN]: 'Endian mismatch',
    [LibbpfErrno.INTERNAL]: 'Internal error in libbpf',
    [LibbpfErrno.RELOC]: 'Relocation failed',
    [LibbpfErrno.LOAD]: 'Load program failure for unknown reason',
    [LibbpfErrno.VERIFY]: 'Kernel verifier blocks program loading',
    [LibbpfErrno.PROG2BIG]: 'Program too big',
    [LibbpfErrno.KVER]: 'Incorrect kernel version',
    [LibbpfErrno.PROGTYPE]: "Kernel doesn't support this program type",
    [LibbpfErrno.WRNGPID]: 'Wrong pid in netlink message',
    [LibbpfErrno.INVSEQ]: 'Invalid netlink sequence',
    [LibbpfErrno.NLPARSE]: 'netlink parsing error',
};
class BPFError extends Error {
    constructor(errno, operation, count) {
        const code = BPFError.getCode(errno);
        const message = Object.hasOwnProperty.call(exports.libbpfErrnoMessages, errno) ? exports.libbpfErrnoMessages[errno] : null;
        super(message || code);
        this.name = 'BPFError';
        this.operation = operation;
        this.errno = errno;
        this.code = code;
        if (count !== undefined)
            this.count = count;
    }
    static getCode(errno) {
        // nonstandard IIRC, but this is how it's called in kernel code
        if (errno === 524)
            return 'ENOTSUPP';
        if (Object.hasOwnProperty.call(LibbpfErrno, errno))
            return LibbpfErrno[errno];
        return getSystemErrorName(-errno);
    }
}
exports.BPFError = BPFError;
function checkStatus(operation, status, count) {
    if (status < 0)
        throw new BPFError(-status, operation, count);
}
exports.checkStatus = checkStatus;
